rafthttp/stream.go

```diff
@@ -0,0 +1,353 @@
+package rafthttp
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/etcdserver/stats"
+	"github.com/coreos/etcd/pkg/types"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+type streamType string
+
+const (
+	streamTypeGeneral streamType = "general"
+	streamTypeMsgApp  streamType = "msgapp"
+
+	streamBufSize = 4096
+)
+
+var (
+	// linkHeartbeatMessage is a special message used as heartbeat message in
+	// link layer. It never conflicts with messages from raft because raft
+	// doesn't send out messages without From and To fields.
+	linkHeartbeatMessage = raftpb.Message{Type: raftpb.MsgHeartbeat}
+)
+
+func isLinkHeartbeatMessage(m raftpb.Message) bool {
+	return m.Type == raftpb.MsgHeartbeat && m.From == 0 && m.To == 0
+}
+
+type outgoingConn struct {
+	t       streamType
+	termStr string
+	io.Writer
+	http.Flusher
+	io.Closer
+}
+
+// streamWriter is a long-running worker that writes messages into the
+// attached outgoingConn.
+type streamWriter struct {
+	fs *stats.FollowerStats
+
+	mu      sync.Mutex // guard field working
+	working bool
+
+	msgc  chan raftpb.Message
+	connc chan *outgoingConn
+	stopc chan struct{}
+	done  chan struct{}
+}
+
+func startStreamWriter(fs *stats.FollowerStats) *streamWriter {
+	w := &streamWriter{
+		fs:    fs,
+		msgc:  make(chan raftpb.Message, streamBufSize),
+		connc: make(chan *outgoingConn),
+		stopc: make(chan struct{}),
+		done:  make(chan struct{}),
+	}
+	go w.run()
+	return w
+}
+
+func (cw *streamWriter) run() {
+	var msgc chan raftpb.Message
+	var heartbeatc <-chan time.Time
+	var t streamType
+	var msgAppTerm uint64
+	var enc encoder
+	var flusher http.Flusher
+	var closer io.Closer
+	tickc := time.Tick(ConnReadTimeout / 3)
+	reset := func() {
```

Comment:
basically, you want to use a method on a receiver to implement an interface or avoid name collisions. When you are not sure or it creates inconvenience, use normal function calls.